dist/index.d.ts

```diff
@@ -4144,7 +4161,27 @@ interface SpawnOptions {
     directions?: DirectionConstant[];
 }
 
-interface SpawningConstructor extends _Constructor<Spawning>, _ConstructorById<Spawning> { }
+interface SpawningConstructor extends _Constructor<Spawning>, _ConstructorById<Spawning> {}
+interface Store extends Record<ResourceConstant, number> {
+    readonly prototype: Store;
+    /**
+     * Returns capacity of this store for the specified resource, or total capacity if resource is undefined.
+     * @param resource The type of the resource
+     */
+    getCapacity(resource?: ResourceConstant): number | null;
+
+    /**
+     * A shorthand for getCapacity(resource) - getUsedCapacity(resource)
+     * @param resource The type of the resource
+     */
+    getFreeCapacity(resource?: ResourceConstant): number | null;
+
+    /**
+     * Returns the capacity used by the specified resource, or total used capacity for general purpose stores if resource is undefined.
+     * @param resource The type of the resource.
+     */
+    getUsedCapacity(resource?: ResourceConstant): number | null;
+}
```

Comment:
This is nice since it can be used for any Store, and allows some code to treat all stores in a generic way.

However, when dealing with a specific structure, like an Extension, it would be nice to be able to do `extension.store.getFreeCapacity(RESOURCE_ENERGY)` without having to check for `null` just after (as we know that for an extension, it will never be null).

For now locally I use this more complex definitions, but which allows to be as close as possible to reality:
```
/** A general purpose Store, which has a limited capacity */
type StoreDefinition = Store<ResourceConstant, false>;

/** A general purpose Store, which has an unlimited capacity */
type StoreDefinitionUnlimited = Store<ResourceConstant, true>;

/** A Store which can only contains some types of resources */
type Store<POSSIBLE_RESSOURCES extends ResourceConstant, UNLIMITED_STORE extends boolean> = {
    /** Returns capacity of this store for the specified resource, or total capacity if resource is undefined. */
    getCapacity<R extends ResourceConstant | undefined>(resource?: R): UNLIMITED_STORE extends true ? null : (undefined extends R ? (ResourceConstant extends POSSIBLE_RESSOURCES ? number : null) : (R extends POSSIBLE_RESSOURCES ? number : null));
    /** Returns the capacity used by the specified resource, or total used capacity for general purpose stores if resource is undefined. */
    getUsedCapacity<R extends ResourceConstant | undefined>(resource?: R): undefined extends R ? (ResourceConstant extends POSSIBLE_RESSOURCES ? number : null) : (R extends POSSIBLE_RESSOURCES ? number : 0);
    /** A shorthand for getCapacity(resource) - getUsedCapacity(resource). */
    getFreeCapacity(resource?: ResourceConstant): number;
}
    & { [P in POSSIBLE_RESSOURCES]: number }
    & { [P in Exclude<ResourceConstant, POSSIBLE_RESSOURCES>]: 0 };

type GenericStore = {
    /** Returns capacity of this store for the specified resource, or total capacity if resource is undefined. */
    getCapacity(resource?: ResourceConstant): number | null;
    /** Returns the capacity used by the specified resource, or total used capacity for general purpose stores if resource is undefined. */
    getUsedCapacity(resource?: ResourceConstant): number | null;
    /** A shorthand for getCapacity(resource) - getUsedCapacity(resource). */
    getFreeCapacity(resource?: ResourceConstant): number;
}
    & { [P in ResourceConstant]: number };
```

With this, for example the StructurePowerSpawn has a store as such:
`store: Store<RESOURCE_ENERGY | RESOURCE_GHODIUM, false>;`
This way when using a StructurePowerSpawn, we can use it in a perfectly typed way.
And to write generic code, you can still assign any specific Store<?> to GenericStore, to have the best of two worlds.

What do you think?

Reply:
This issue has also been discussed a bit on Slack. It's a bit unfortunate the game seemed to give the impression that since everything now has a `.store` that they would all be consistent. That not being the case, I think the more complex typing is good to we don't need null checks all over the place.

Reply:
Will implement the change later today and update the PR.

Reply:
I tried to apply the suggest changes.

While doing so I needed to disable the StoreDefinition test as this did not work because of the way the store is defined.

Could someone help me to recreate the test for the new store definition ? (This might indicate a breaking change)

Or should I just add a new test instead of the existing one ?

Reply:
Oh, I see, the test leverages an helper function `keys`, which now sees the newly added methods (getCapacity, getUsedCapacity and getFreeCapacity).
I don't have a perfect solution, but as a workaround another helper function could be used instead:
```
function resources(o: GenericStore): Array<ResourceConstant> {
    return Object.keys(o) as Array<ResourceConstant>;
}
```
Then the test will pass again when using `resources` instead of `keys`